Race between layer and Lambda update

The example below is from `anvildev`, however the same errors occurred on `dev` and `anvilprod`.

During the deploy [job](https://gitlab.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/33582) for the merge commit of PR #5909 (which updated the `elasticsearch` client from `7.10.1` to `7.17.9`), an `UnsupportedProductError` occurred for both the `indexercachehealth` and `servicecachehealth` lambdas.

[CloudWatch Insights logs](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:logs-insights$3FqueryDetail$3D~(end~'2024-02-03T06*3a24*3a53.173Z~start~'2024-02-03T06*3a14*3a41.347Z~timeType~'ABSOLUTE~tz~'UTC~editorString~'fields*20*40timestamp*2c*20*40log*2c*20*40message*0a*7c*20filter*20*40message*20like*20*2fERROR*7cTask*2f*0a*23*20*7c*20filter*20*40requestId*20like*20*2f*2f*0a*7c*20sort*20*40timestamp*20asc*0a*7c*20limit*20200~queryId~'7d5afde2-5bee-40a0-ac1c-cecc296f4944~source~(~'*2faws*2flambda*2fazul-indexer-anvildev-indexercachehealth~'*2faws*2flambda*2fazul-service-anvildev-servicecachehealth))):

```
[ERROR] UnsupportedProductError: The client noticed that the server is not a supported distribution of Elasticsearch
Traceback (most recent call last):
  File "/var/task/azul/chalice.py", line 166, in patched_event_source_handler
    return old_handler(self_, event, context)
  File "/var/task/chalice/app.py", line 1756, in __call__
    return self.handler(event_obj)
  File "/var/task/app.py", line 212, in update_health_cache
    app.health_controller.update_cache()
  File "/var/task/azul/health.py", line 138, in update_cache
    health_object = dict(time=time.time(), health=self._health.as_json_fast())
  File "/var/task/azul/health.py", line 308, in as_json_fast
    return self.as_json(p.key for p in self.fast_properties[self.lambda_name])
  File "/var/task/azul/health.py", line 181, in as_json
    json = {k: getattr(self, k) for k in sorted(keys)}
  File "/var/task/azul/health.py", line 181, in <dictcomp>
    json = {k: getattr(self, k) for k in sorted(keys)}
  File "/var/task/azul/health.py", line 73, in __get__
    return super().__get__(obj, objtype=objtype)
  File "/var/task/azul/caching.py", line 189, in __get__
    value = obj.__dict__[self.fget.__name__] = self.fget(obj)
  File "/var/task/azul/health.py", line 273, in elasticsearch
    'up': ESClientFactory.get().ping(),
  File "/opt/python/elasticsearch/client/utils.py", line 347, in _wrapped
    return func(*args, params=params, headers=headers, **kwargs)
  File "/opt/python/elasticsearch/client/__init__.py", line 280, in ping
    return self.transport.perform_request(
  File "/opt/python/elasticsearch/transport.py", line 421, in perform_request
    _ProductChecker.raise_error(self._verified_elasticsearch)
  File "/opt/python/elasticsearch/transport.py", line 638, in raise_error
    raise UnsupportedProductError(message)
```

> Spike to provide more reliable reproduction

Reproduction:

- Check out `develop`
- Apply patch 1 (adds `/test` endpoint and `ua_parser` requirement)
- Run `make requirements_update` to create wheel for `ua_parser`
- Deploy to personal deployment
- Revert changes from patch 1 and delete `ua_parser` wheel
- Apply patch 2 (keeps `/test` endpoint but has no `ua_parser` requirement)
- Deploy to personal deployment

Approximately 16 seconds after the last deploy was started, there was a window of about 20 seconds where the Lambda function is using the old code (with the `ua_parser` requirement) however the Lambda layer had already been updated to not contain `ua_parser`.


## azul-trail-dev    

| eventTime | eventType | eventName | requestParameters.functionName |
| --- | --- | --- | --- |
| 2025-09-17T23:53:33Z | AwsApiCall | UpdateFunctionConfiguration20150331v2 | azul-service-daniel-manifest |
| 2025-09-17T23:53:34Z | AwsApiCall | UpdateFunctionConfiguration20150331v2 | azul-service-daniel |
| 2025-09-17T23:53:34Z | AwsApiCall | UpdateFunctionConfiguration20150331v2 | azul-service-daniel-servicecachehealth |
| 2025-09-17T23:53:41Z | AwsApiCall | UpdateFunctionCode20150331v2 | azul-service-daniel-manifest |
| 2025-09-17T23:53:49Z | AwsApiCall | UpdateFunctionCode20150331v2 | azul-service-daniel-servicecachehealth |
| 2025-09-17T23:53:56Z | AwsApiCall | UpdateFunctionCode20150331v2 | azul-service-daniel |

## Pinging the `/test` endpoint once every second
(started a fraction of a second after the 2nd deploy was started)

```
$ python scripts/request_flooder.py --method GET --url https://service.daniel.dev.singlecell.gi.ucsc.edu/test --rate 1 --per 1
2025-09-17 16:53:23,649    INFO MainThread __main__: Starting requests at a rate of 1 requests per 1 seconds for 300 seconds
2025-09-17 16:53:24,654    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:25,315    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.660s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:25,658    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:25,874    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.216s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:26,663    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:26,848    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.185s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:27,667    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:27,965    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.298s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:28,671    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:28,838    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.167s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:29,673    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:29,848    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.175s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:30,673    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:30,851    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.178s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:31,678    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:31,850    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.173s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:32,682    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:32,861    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.178s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:33,687    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:33,850    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.163s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:34,691    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:34,868    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.177s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:35,696    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:35,871    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.175s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:36,696    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:36,992    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 0.296s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:37,701    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:38,707    INFO ThreadPoolExecutor-0_1 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:39,711    INFO ThreadPoolExecutor-0_2 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:40,630    INFO ThreadPoolExecutor-0_0 __main__: Got 500 response after 2.928s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:40,714    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:40,878    INFO ThreadPoolExecutor-0_0 __main__: Got 500 response after 0.164s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:41,306    INFO ThreadPoolExecutor-0_1 __main__: Got 500 response after 2.599s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:41,719    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:41,890    INFO ThreadPoolExecutor-0_0 __main__: Got 500 response after 0.171s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:42,714    INFO ThreadPoolExecutor-0_2 __main__: Got 500 response after 3.002s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:42,724    INFO ThreadPoolExecutor-0_1 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:42,890    INFO ThreadPoolExecutor-0_1 __main__: Got 500 response after 0.166s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:43,728    INFO ThreadPoolExecutor-0_2 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:43,896    INFO ThreadPoolExecutor-0_2 __main__: Got 500 response after 0.169s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:44,733    INFO ThreadPoolExecutor-0_1 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:45,033    INFO ThreadPoolExecutor-0_1 __main__: Got 500 response after 0.300s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:45,738    INFO ThreadPoolExecutor-0_2 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:45,911    INFO ThreadPoolExecutor-0_2 __main__: Got 500 response after 0.172s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:46,743    INFO ThreadPoolExecutor-0_1 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:47,031    INFO ThreadPoolExecutor-0_1 __main__: Got 500 response after 0.287s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:47,747    INFO ThreadPoolExecutor-0_2 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:47,900    INFO ThreadPoolExecutor-0_2 __main__: Got 500 response after 0.152s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:48,753    INFO ThreadPoolExecutor-0_1 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:48,931    INFO ThreadPoolExecutor-0_1 __main__: Got 500 response after 0.179s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:49,758    INFO ThreadPoolExecutor-0_2 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:49,924    INFO ThreadPoolExecutor-0_2 __main__: Got 500 response after 0.166s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:50,762    INFO ThreadPoolExecutor-0_1 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:51,051    INFO ThreadPoolExecutor-0_1 __main__: Got 500 response after 0.289s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:51,767    INFO ThreadPoolExecutor-0_2 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:51,938    INFO ThreadPoolExecutor-0_2 __main__: Got 500 response after 0.170s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:52,770    INFO ThreadPoolExecutor-0_1 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:52,938    INFO ThreadPoolExecutor-0_1 __main__: Got 500 response after 0.168s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:53,775    INFO ThreadPoolExecutor-0_2 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:53,949    INFO ThreadPoolExecutor-0_2 __main__: Got 500 response after 0.174s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:54,778    INFO ThreadPoolExecutor-0_1 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:55,061    INFO ThreadPoolExecutor-0_1 __main__: Got 500 response after 0.283s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:55,784    INFO ThreadPoolExecutor-0_2 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:55,954    INFO ThreadPoolExecutor-0_2 __main__: Got 500 response after 0.170s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:56,787    INFO ThreadPoolExecutor-0_1 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:56,962    INFO ThreadPoolExecutor-0_1 __main__: Got 500 response after 0.174s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:57,789    INFO ThreadPoolExecutor-0_2 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:58,110    INFO ThreadPoolExecutor-0_2 __main__: Got 500 response after 0.321s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:58,792    INFO ThreadPoolExecutor-0_1 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:53:59,088    INFO ThreadPoolExecutor-0_1 __main__: Got 500 response after 0.295s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:53:59,797    INFO ThreadPoolExecutor-0_2 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:54:00,801    INFO ThreadPoolExecutor-0_1 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:54:01,806    INFO ThreadPoolExecutor-0_0 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:54:02,613    INFO ThreadPoolExecutor-0_2 __main__: Got 200 response after 2.815s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:54:02,811    INFO ThreadPoolExecutor-0_2 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:54:02,975    INFO ThreadPoolExecutor-0_2 __main__: Got 200 response after 0.164s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:54:03,585    INFO ThreadPoolExecutor-0_1 __main__: Got 200 response after 2.784s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:54:03,816    INFO ThreadPoolExecutor-0_2 __main__: Making GET request to 'https://service.daniel.dev.singlecell.gi.ucsc.edu/test'
2025-09-17 16:54:03,987    INFO ThreadPoolExecutor-0_2 __main__: Got 200 response after 0.171s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
2025-09-17 16:54:04,359    INFO ThreadPoolExecutor-0_0 __main__: Got 200 response after 2.553s from GET to https://service.daniel.dev.singlecell.gi.ucsc.edu/test
...
```


### /aws/lambda/azul-service-daniel

```
[ERROR] 2025-09-17T23:53:40.635Z    58e5aae6-3b96-49f1-b5b6-929c5c3293a5    azul-service-daniel Caught exception for path /test
Traceback (most recent call last):
  File "/var/task/chalice/app.py", line 1922, in _get_view_function_response
    response = view_function(**function_args)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/app.py", line 467, in parse_user_agent
    import ua_parser
ModuleNotFoundError: No module named 'ua_parser'
```

### patch 1

```diff
Index: requirements.txt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/requirements.txt b/requirements.txt
--- a/requirements.txt  (revision bd954793e19668ef7f19f2fe3f7d4fe08116cce2)
+++ b/requirements.txt  (date 1758153418316)
@@ -20,6 +20,7 @@
 resumablehash==1.2  # match version with --find-links above
 rsa==4.7.2  # < 4.8, see https://github.com/DataBiosphere/azul/issues/6774. Resolve build-time dependency ambguity
 setuptools==80.9.0  # Match requirements.pip.txt. Python 3.12 removed distutils, which we depended on transitively. Luckily, setuptools includes a vendored copy.
+ua-parser[regex]==1.0.1
 urllib3==1.26.20 # < 2, see https://github.com/DataBiosphere/azul/issues/6777, match with types-urllib in requirements.dev.txt
 werkzeug==3.1.3
 -r requirements.trans.txt
Index: lambdas/service/app.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lambdas/service/app.py b/lambdas/service/app.py
--- a/lambdas/service/app.py    (revision bd954793e19668ef7f19f2fe3f7d4fe08116cce2)
+++ b/lambdas/service/app.py    (date 1758153418315)
@@ -454,6 +454,24 @@
 globals().update(app.default_routes())
 
 
+@app.route(
+    '/test',
+    methods=['GET'],
+    cors=True,
+    spec={
+        'summary': 'Test endpoint',
+        'tags': ['Test'],
+    }
+)
+def parse_user_agent():
+    import ua_parser
+    ua_string = app.current_request.headers.get('User-Agent')
+    parsed = ua_parser.parse(ua_string)
+    return {
+        'ua': str(parsed)
+    }
+
+
 @app.route(
     '/oauth2_redirect',
     enabled=config.google_oauth2_client_id is not None,
```

### patch 2

```diff

Index: lambdas/service/app.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lambdas/service/app.py b/lambdas/service/app.py
--- a/lambdas/service/app.py    (revision bd954793e19668ef7f19f2fe3f7d4fe08116cce2)
+++ b/lambdas/service/app.py    (date 1758152884852)
@@ -454,6 +454,21 @@
 globals().update(app.default_routes())
 
 
+@app.route(
+    '/test',
+    methods=['GET'],
+    cors=True,
+    spec={
+        'summary': 'Test endpoint',
+        'tags': ['Test'],
+    }
+)
+def parse_user_agent():
+    return {
+        'ua': app.current_request.headers.get('User-Agent')
+    }
+
+
 @app.route(
     '/oauth2_redirect',
     enabled=config.google_oauth2_client_id is not None,
```

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -894,7 +894,7 @@ class Chalice:
             permissions_by_function = defaultdict(set)
             permissions = resources['aws_lambda_permission']
             for permission_name, permission in json_item_dicts(permissions):
-                alias = alias_ref(permission['function_name'])
+                alias = alias_ref(json_str(permission['function_name']))
                 permissions_by_function[alias].add(permission_name)
             permissions_by_function = dict(permissions_by_function)
             for resource_name, notification in resource_items(resource_type):
@@ -902,7 +902,9 @@ class Chalice:
                 notification['depends_on'] = [
                     f'aws_lambda_permission.{permission_name}'
                     for function in json_element_dicts(notification['lambda_function'])
-                    for permission_name in permissions_by_function[function['lambda_function_arn']]
+                    for permission_name in permissions_by_function[json_str(
+                        function['lambda_function_arn']
+                    )]
                 ]
         else:
             assert resource_type not in resources

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -51,7 +51,6 @@ from azul.types import (
     json_element_dicts,
     json_item_dicts,
     json_item_mappings,
-    json_list_of_dicts,
     json_mapping,
     json_str,
     not_none,
@@ -788,40 +787,66 @@ class Chalice:
         # Replace any references to unqualified function ARNs in the OpenAPI
         # spec emitted by Chalice with references to the alias.
         #
-        def functions_to_aliases[T: AnyMutableJSON](v: T) -> T:
+        def alias_ref(v: str) -> str:
+            return v.replace('${aws_lambda_function', '${aws_lambda_alias')
+
+        def alias_invoke_arns[T: AnyMutableJSON](v: T) -> T:
             if isinstance(v, dict):
-                return type(v)((k, functions_to_aliases(v)) for k, v in v.items())
+                return type(v)((k, alias_invoke_arns(v)) for k, v in v.items())
             elif isinstance(v, list):
-                return type(v)(map(functions_to_aliases, v))
+                return type(v)(map(alias_invoke_arns, v))
             elif isinstance(v, str) and v.endswith('.invoke_arn}'):
-                r = function_to_alias(v)
+                r = alias_ref(v)
                 assert isinstance(r, type(v))
                 return r
             else:
                 return v
 
-        def function_to_alias(v: str) -> str:
-            return v.replace('${aws_lambda_function', '${aws_lambda_alias')
-
         openapi_spec = json_dict(json.loads(json_str(locals[app_name])))
-        openapi_spec = functions_to_aliases(openapi_spec)
+        openapi_spec = alias_invoke_arns(openapi_spec)
 
         # Replace any references to unqualified function ARNs in the resources
         # emitted by Chalice with references to the alias.
         #
-        for _, r in json_item_dicts(resources.get('aws_lambda_permission', {})):
-            alias = function_to_alias(json_str(r['function_name']))
+        # There are two ways for a aws_lambda_permission resource to reference a
+        # Lambda function alias: as a suffix, by appending the alias name to the
+        # function ARN in the `function_name` property, or by setting the
+        # `qualifier` property of the aws_lambda_permission resource to the
+        # alias name. Unfortunately, the suffix reference is silently converted
+        # to the `qualifier` property and Terraform treats this conversion as
+        # drift, resulting in perpetually non-empty plans. To avoid the drift,
+        # we use the second method directly.
+        #
+        for resource_name, resource in resource_items('aws_lambda_permission'):
+            alias = alias_ref(json_str(resource['function_name']))
             assert alias.endswith('.arn}'), alias
-            r['qualifier'] = alias.replace('.arn', '.name')
-        for _, r in json_item_dicts(resources.get('aws_cloudwatch_event_target', {})):
-            r['arn'] = function_to_alias(json_str(r['arn']))
-        for _, r in json_item_dicts(resources.get('aws_lambda_event_source_mapping', {})):
-            r['function_name'] = function_to_alias(json_str(r['function_name']))
-        for _, r in json_item_dicts(resources.get('aws_s3_bucket_notification', {})):
-            for lf in json_list_of_dicts(r['lambda_function']):
-                lf['lambda_function_arn'] = function_to_alias(
-                    json_str(lf['lambda_function_arn'])
-                )
+            assert 'qualifier' not in resource
+            resource['qualifier'] = alias.replace('.arn', '.name')
+
+        # Patch any remaining resources with properties holding function ARNs
+        # with the ARN of the corresponding alias.
+        #
+        def alias_property(property_name: str, resource: MutableJSON):
+            alias = alias_ref(json_str(resource[property_name]))
+            resource[property_name] = alias
+
+        for resource_name, resource in resource_items('aws_cloudwatch_event_target'):
+            alias_property('arn', resource)
+
+        resource_type = 'aws_lambda_event_source_mapping'
+        if app_name == 'indexer':
+            for resource_name, resource in resource_items(resource_type):
+                alias_property('function_name', resource)
+        else:
+            assert resource_type not in resources
+
+        resource_type = 'aws_s3_bucket_notification'
+        if config.enable_log_forwarding and app_name == 'indexer':
+            for resource_name, resource in resource_items(resource_type):
+                for notified_function in json_element_dicts(resource['lambda_function']):
+                    alias_property('lambda_function_arn', notified_function)
+        else:
+            assert resource_type not in resources
 
         # Pre-create a CloudWatch log group for every Lambda function so that
         # we can set the log retention explicitly.
@@ -869,8 +894,8 @@ class Chalice:
             permissions_by_function = defaultdict(set)
             permissions = resources['aws_lambda_permission']
             for permission_name, permission in json_item_dicts(permissions):
-                function_ref = permission['function_name']
-                permissions_by_function[function_ref].add(permission_name)
+                alias = alias_ref(permission['function_name'])
+                permissions_by_function[alias].add(permission_name)
             permissions_by_function = dict(permissions_by_function)
             for resource_name, notification in resource_items(resource_type):
                 assert 'depends_on' not in notification, notification

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -891,20 +891,17 @@ class Chalice:
             }
             # To prevent a race condition by Terraform, we make the bucket
             # notifications depend on the related aws_lambda_permission.
-            permissions_by_function = defaultdict(set)
-            permissions = resources['aws_lambda_permission']
-            for permission_name, permission in json_item_dicts(permissions):
+            permissions = defaultdict(set)
+            for permission_name, permission in resource_items('aws_lambda_permission'):
                 alias = alias_ref(json_str(permission['function_name']))
-                permissions_by_function[alias].add(permission_name)
-            permissions_by_function = dict(permissions_by_function)
+                permissions[alias].add(permission_name)
+            permissions = dict(permissions)
             for resource_name, notification in resource_items(resource_type):
                 assert 'depends_on' not in notification, notification
                 notification['depends_on'] = [
                     f'aws_lambda_permission.{permission_name}'
                     for function in json_element_dicts(notification['lambda_function'])
-                    for permission_name in permissions_by_function[json_str(
-                        function['lambda_function_arn']
-                    )]
+                    for permission_name in permissions[json_str(function['lambda_function_arn'])]
                 ]
         else:
             assert resource_type not in resources

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -957,7 +957,8 @@ class Chalice:
         if app_name == 'indexer':
             event_source_mappings = resources['aws_lambda_event_source_mapping']
             for _, resource in json_item_dicts(event_source_mappings):
-                _, _, resource_name = json_str(resource['event_source_arn']).rpartition(':')
+                arn = json_str(resource['event_source_arn'])
+                _, _, resource_name = arn.rpartition(':')
                 suffix = '.fifo' if resource_name.endswith('.fifo') else ''
                 sqs_name, _ = config.unqualified_resource_name(resource_name, suffix)
                 resource['event_source_arn'] = f'${{aws_sqs_queue.{sqs_name}.arn}}'

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -801,6 +801,9 @@ class Chalice:
         openapi_spec = function_to_alias(openapi_spec)
         resources = function_to_alias(resources)
 
+        # Pre-create a CloudWatch log group for every Lambda function so that
+        # we can set the log retention explicitly.
+        #
         assert 'aws_cloudwatch_log_group' not in resources
         functions = json_item_dicts(resources['aws_lambda_function'])
         resources['aws_cloudwatch_log_group'] = {

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -834,12 +834,11 @@ class Chalice:
         # (https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html),
         # so replacing the periods with underscores results in valid resource
         # names while retaining the correlation with bucket names.
-        try:
-            bucket_notifications = resources['aws_s3_bucket_notification']
-        except KeyError:
-            pass
-        else:
-            resources['aws_s3_bucket_notification'] = {
+        #
+        resource_type = 'aws_s3_bucket_notification'
+        if config.enable_log_forwarding and app_name == 'indexer':
+            bucket_notifications = resources[resource_type]
+            resources[resource_type] = {
                 key.replace('.', '_'): value
                 for key, value in json_item_dicts(bucket_notifications)
             }
@@ -850,13 +849,16 @@ class Chalice:
             for permission_name, permission in json_item_dicts(permissions):
                 function_ref = permission['function_name']
                 permissions_by_function[function_ref].add(permission_name)
-            for _, notification in json_item_dicts(resources['aws_s3_bucket_notification']):
+            permissions_by_function = dict(permissions_by_function)
+            for _, notification in json_item_dicts(resources[resource_type]):
                 assert 'depends_on' not in notification, notification
                 notification['depends_on'] = [
                     f'aws_lambda_permission.{permission_name}'
                     for function in json_element_dicts(notification['lambda_function'])
                     for permission_name in permissions_by_function[function['lambda_function_arn']]
                 ]
+        else:
+            assert resource_type not in resources
 
         # The fix for https://github.com/aws/chalice/issues/1237 introduced the
         # create_before_destroy hack and it may have helped but has far-ranging

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -743,6 +743,9 @@ class Chalice:
         data = json_dict(tf_config['data'])
         locals = json_dict(tf_config['locals'])
 
+        def resource_items(resource_type: str) -> Iterable[tuple[str, MutableJSON]]:
+            return json_item_dicts(resources[resource_type])
+
         # null_data_source has been deprecated and locals should be used instead.
         # However, the data sources defined underneath it aren't actually used
         # anywhere so we can just delete the entry.
@@ -757,8 +760,7 @@ class Chalice:
                 '${aws_vpc_endpoint.%s.id}' % app_name
             ]
 
-        functions = json_item_dicts(resources['aws_lambda_function'])
-        for resource_name, resource in functions:
+        for resource_name, resource in resource_items('aws_lambda_function'):
             assert 'layers' not in resource
             resource['layers'] = ['${aws_lambda_layer_version.dependencies.arn}']
             # Publishing a new Lambda function version each time lets us perform
@@ -805,13 +807,12 @@ class Chalice:
         # we can set the log retention explicitly.
         #
         assert 'aws_cloudwatch_log_group' not in resources
-        functions = json_item_dicts(resources['aws_lambda_function'])
         resources['aws_cloudwatch_log_group'] = {
             f'{resource_name}_lambda': {
                 'name': f'/aws/lambda/{resource['function_name']}',
                 'retention_in_days': config.audit_log_retention_days
             }
-            for resource_name, resource in functions
+            for resource_name, resource in resource_items('aws_lambda_function')
         }
 
         for resource_type, property_name in [
@@ -822,7 +823,7 @@ class Chalice:
             # need them to be prefixed with `azul-` to allow for limiting the
             # scope of certain IAM permissions for Gitlab and, more importantly,
             # the deployment stage so these resources are segregated by deployment.
-            for _, resource in json_item_dicts(resources[resource_type]):
+            for resource_name, resource in resource_items(resource_type):
                 unqualified = json_str(resource[property_name])
                 function_name, _, suffix = unqualified.partition('-')
                 assert suffix == 'event', suffix
@@ -851,7 +852,7 @@ class Chalice:
                 function_ref = permission['function_name']
                 permissions_by_function[function_ref].add(permission_name)
             permissions_by_function = dict(permissions_by_function)
-            for _, notification in json_item_dicts(resources[resource_type]):
+            for resource_name, notification in resource_items(resource_type):
                 assert 'depends_on' not in notification, notification
                 notification['depends_on'] = [
                     f'aws_lambda_permission.{permission_name}'
@@ -956,7 +957,7 @@ class Chalice:
         #
         if app_name == 'indexer':
             event_source_mappings = resources['aws_lambda_event_source_mapping']
-            for _, resource in json_item_dicts(event_source_mappings):
+            for resource_name, resource in json_item_dicts(event_source_mappings):
                 arn = json_str(resource['event_source_arn'])
                 _, _, resource_name = arn.rpartition(':')
                 suffix = '.fifo' if resource_name.endswith('.fifo') else ''
@@ -967,7 +968,7 @@ class Chalice:
         # deleted until after the permissions for the new aliases have been
         # created.
         #
-        for name, resource in json_item_dicts(resources['aws_lambda_permission']):
+        for name, resource in resource_items('aws_lambda_permission'):
             assert 'lifecycle' not in resource, resource
             resource['lifecycle'] = {'create_before_destroy': True}
 

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -823,7 +823,8 @@ class Chalice:
             # scope of certain IAM permissions for Gitlab and, more importantly,
             # the deployment stage so these resources are segregated by deployment.
             for _, resource in json_item_dicts(resources[resource_type]):
-                function_name, _, suffix = json_str(resource[property_name]).partition('-')
+                unqualified = json_str(resource[property_name])
+                function_name, _, suffix = unqualified.partition('-')
                 assert suffix == 'event', suffix
                 assert function_name, function_name
                 resource[property_name] = config.qualified_resource_name(function_name)

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -51,6 +51,7 @@ from azul.types import (
     json_element_dicts,
     json_item_dicts,
     json_item_mappings,
+    json_list_of_dicts,
     json_mapping,
     json_str,
     not_none,
@@ -784,24 +785,43 @@ class Chalice:
             resource['source_code_hash'] = '${filebase64sha256("%s")}' % package_zip
             resource['filename'] = package_zip
 
-        # Replace any references to unqualified function ARNs emitted by Chalice
-        # with references to the alias.
+        # Replace any references to unqualified function ARNs in the OpenAPI
+        # spec emitted by Chalice with references to the alias.
         #
-        def function_to_alias[T: AnyMutableJSON](v: T) -> T:
+        def functions_to_aliases[T: AnyMutableJSON](v: T) -> T:
             if isinstance(v, dict):
-                return type(v)((k, function_to_alias(v)) for k, v in v.items())
+                return type(v)((k, functions_to_aliases(v)) for k, v in v.items())
             elif isinstance(v, list):
-                return type(v)(map(function_to_alias, v))
-            elif isinstance(v, str) and v.endswith(('.arn}', '.invoke_arn}')):
-                r = v.replace('${aws_lambda_function', '${aws_lambda_alias')
+                return type(v)(map(functions_to_aliases, v))
+            elif isinstance(v, str) and v.endswith('.invoke_arn}'):
+                r = function_to_alias(v)
                 assert isinstance(r, type(v))
                 return r
             else:
                 return v
 
+        def function_to_alias(v: str) -> str:
+            return v.replace('${aws_lambda_function', '${aws_lambda_alias')
+
         openapi_spec = json_dict(json.loads(json_str(locals[app_name])))
-        openapi_spec = function_to_alias(openapi_spec)
-        resources = function_to_alias(resources)
+        openapi_spec = functions_to_aliases(openapi_spec)
+
+        # Replace any references to unqualified function ARNs in the resources
+        # emitted by Chalice with references to the alias.
+        #
+        for _, r in json_item_dicts(resources.get('aws_lambda_permission', {})):
+            alias = function_to_alias(json_str(r['function_name']))
+            assert alias.endswith('.arn}'), alias
+            r['qualifier'] = alias.replace('.arn', '.name')
+        for _, r in json_item_dicts(resources.get('aws_cloudwatch_event_target', {})):
+            r['arn'] = function_to_alias(json_str(r['arn']))
+        for _, r in json_item_dicts(resources.get('aws_lambda_event_source_mapping', {})):
+            r['function_name'] = function_to_alias(json_str(r['function_name']))
+        for _, r in json_item_dicts(resources.get('aws_s3_bucket_notification', {})):
+            for lf in json_list_of_dicts(r['lambda_function']):
+                lf['lambda_function_arn'] = function_to_alias(
+                    json_str(lf['lambda_function_arn'])
+                )
 
         # Pre-create a CloudWatch log group for every Lambda function so that
         # we can set the log retention explicitly.

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -969,7 +969,7 @@ class Chalice:
         # created.
         #
         for resource_name, resource in resource_items('aws_lambda_permission'):
-            assert 'lifecycle' not in resource, resource
+            assert 'lifecycle' not in resource, (resource_name, resource)
             resource['lifecycle'] = {'create_before_destroy': True}
 
         return {

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -968,7 +968,7 @@ class Chalice:
         # deleted until after the permissions for the new aliases have been
         # created.
         #
-        for name, resource in resource_items('aws_lambda_permission'):
+        for resource_name, resource in resource_items('aws_lambda_permission'):
             assert 'lifecycle' not in resource, resource
             resource['lifecycle'] = {'create_before_destroy': True}
 

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -814,7 +814,7 @@ class Chalice:
             for resource_name, resource in functions
         }
 
-        for resource_type, argument in [
+        for resource_type, property_name in [
             ('aws_cloudwatch_event_rule', 'name'),
             ('aws_cloudwatch_event_target', 'target_id')
         ]:
@@ -823,10 +823,10 @@ class Chalice:
             # scope of certain IAM permissions for Gitlab and, more importantly,
             # the deployment stage so these resources are segregated by deployment.
             for _, resource in json_item_dicts(resources[resource_type]):
-                function_name, _, suffix = json_str(resource[argument]).partition('-')
+                function_name, _, suffix = json_str(resource[property_name]).partition('-')
                 assert suffix == 'event', suffix
                 assert function_name, function_name
-                resource[argument] = config.qualified_resource_name(function_name)
+                resource[property_name] = config.qualified_resource_name(function_name)
 
         # Chalice-generated S3 bucket notifications include the bucket name in
         # the resource name, resulting in an invalid resource name when the

